# Antes de iniciar qualquer projeto

###  #entenda a necessiade

Este é um projeto de análise de dados que contem fontes de arquivos locais e com estrutura de dados comuns, ou seja, string, float e integer. A minha necessidade no momento é utilizar ferramentas para tratar e explorar dados e chegar a um resumo do censo americano.

Esclarecendo nossas necessidades podemos ter uma ideia de quais recursos precisamos.

### #documentações oficiais das biblotecas

Parte do trabalho de um cientista de dados é explorar dados, podemos dizer que existem bibliotecas (conjuntos de ferramentas), com funções desenvolvidas que serão bastente uteis no dia-a-dia, segue a documentação dos recursos que estarei utilizando:

- <b>random</b>: https://docs.python.org/3/library/random.html
- <b>numpy</b>: https://numpy.org/
- <b>pandas</b>: https://pandas.pydata.org/


### #ao decorrer dos exemplos irei implementar o uso de algumas bibliotecas bastante utilizadas para métodos estatísticos
- <b>matplotlib</b>: https://matplotlib.org/stable/tutorials/index.html
- <b>sklearn</b>: https://scikit-learn.org/stable/install.html


In [1]:
import random
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
# leitura do arquivo .csv representando o census americano
dataset = pd.read_csv("base/census.csv")
dataset

age          workclass  final-weight    education  education-num  \
0       39          State-gov         77516    Bachelors             13   
1       50   Self-emp-not-inc         83311    Bachelors             13   
2       38            Private        215646      HS-grad              9   
3       53            Private        234721         11th              7   
4       28            Private        338409    Bachelors             13   
...    ...                ...           ...          ...            ...   
32556   27            Private        257302   Assoc-acdm             12   
32557   40            Private        154374      HS-grad              9   
32558   58            Private        151910      HS-grad              9   
32559   22            Private        201490      HS-grad              9   
32560   52       Self-emp-inc        287927      HS-grad              9   

            marital-status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital-gain  capital-loos  hour-per-week  native-country  \
0         Male          2174             0             40   United-States   
1         Male             0             0             13   United-States   
2         Male             0             0             40   United-States   
3         Male             0             0             40   United-States   
4       Female             0             0             40            Cuba   
...        ...           ...           ...            ...             ...   
32556   Female             0             0             38   United-States   
32557     Male             0             0             40   United-States   
32558   Female             0             0             40   United-States   
32559     Male             0             0             20   United-States   
32560   Female         15024             0             40   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 15 columns]

# #resumo da base, o comando shape imprime o total de linhas e colunas da base

In [3]:
dataset.shape
print(f"shape: {dataset.shape}")
print(f"linhas: {dataset.shape[0]} | colunas: {dataset.shape[1]}")

shape: (32561, 15)
linhas: 32561 | colunas: 15


# #use o comando head para imprimir os primeiros registros da base

In [4]:
dataset.head()

age          workclass  final-weight   education  education-num  \
0   39          State-gov         77516   Bachelors             13   
1   50   Self-emp-not-inc         83311   Bachelors             13   
2   38            Private        215646     HS-grad              9   
3   53            Private        234721        11th              7   
4   28            Private        338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loos  hour-per-week  native-country  income  
0          2174             0             40   United-States   <=50K  
1             0             0             13   United-States   <=50K  
2             0             0             40   United-States   <=50K  
3             0             0             40   United-States   <=50K  
4             0             0             40            Cuba   <=50K

# #use o comando tail para imprimir os últimos registros da base

In [5]:
dataset.tail()

age      workclass  final-weight    education  education-num  \
32556   27        Private        257302   Assoc-acdm             12   
32557   40        Private        154374      HS-grad              9   
32558   58        Private        151910      HS-grad              9   
32559   22        Private        201490      HS-grad              9   
32560   52   Self-emp-inc        287927      HS-grad              9   

            marital-status          occupation relationship    race      sex  \
32556   Married-civ-spouse        Tech-support         Wife   White   Female   
32557   Married-civ-spouse   Machine-op-inspct      Husband   White     Male   
32558              Widowed        Adm-clerical    Unmarried   White   Female   
32559        Never-married        Adm-clerical    Own-child   White     Male   
32560   Married-civ-spouse     Exec-managerial         Wife   White   Female   

       capital-gain  capital-loos  hour-per-week  native-country  income  
32556             0             0             38   United-States   <=50K  
32557             0             0             40   United-States    >50K  
32558             0             0             40   United-States   <=50K  
32559             0             0             20   United-States   <=50K  
32560         15024             0             40   United-States    >50K

# ----------------------------------------------------------------------

# Amostragem Aleatória Simples

### sample
A função <b>"sample"</b> retorna valores aleatórios de um dataset, alguns parâmetros podem ser passados para controlar a criação de amostras de acordo com a necessidade, como exemplo:  <b>"n"</b>, <b>"replace"</b> e <b>"random_state"</b>.

### #replace
Por padrão esse parâmetro tem o valor <b>False</b>, se alterado para <b>True</b> retorna indices únicos na amostra aleatória.
<br>replace = True --> não repete os mesmos indices.

### #random_state
Caso este parametro receba um valor >= 1 será mantido os mesmos indices na amostra aleatória.
<br>random_state = 1 --> mantem sempre os mesmo indices.

- <b>dataset</b>: a base com os dados;
- <b>n (amostra)</b>: será a quantidade de amostra que desejamos;
- <b>replace</b>: definição se haverá indices exclusivos na amostra;
- <b>random_state</b>: define se queremos a repetição dos indices todas as vezes que gerarmos amostras.




# Atente-se na diferença!
Perceba a difereça entre repetição e reutilização.
- <b>replace</b>: permite/não permite os mesmos indices na amostra, por exemplo, se True não haverá repetição do indice 1;
- <b>random_state</b>: permite/não permite que sejam gerados novos indices, se >=1 será reutilizado os indices anteriores;



Esses são apenas alguns parâmentros, para saber mais acesse a documentação do Pandas:
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sample.html


# Evitando a repetição de código

Seguindo as <b>boas práticas de codificação</b>, irei implementar algumas funções para reutilizar durante a análise e deixar o código mais legível.

Primeiramente irei criar uma função para gerar uma amostra aleatória simples usando os parâmetros <b>n, replace e random_state</b>.<br><br>


In [6]:
def amostragem_aleatoria_simples(dataset, amostras, replace, random_state):
    return dataset.sample(n=amostras, replace=replace, random_state=random_state)

In [7]:
df_amostra_aleatoria_simples = amostragem_aleatoria_simples(dataset, 100, True, 1)
df_amostra_aleatoria_simples

age          workclass  final-weight      education  education-num  \
29733   33            Private         59083      Bachelors             13   
235     59          State-gov        268700        HS-grad              9   
12172   27            Private        104457      Bachelors             13   
5192    32            Private        207668      Bachelors             13   
32511   25          Local-gov        514716      Bachelors             13   
...    ...                ...           ...            ...            ...   
17534   43            Private         54611        HS-grad              9   
9495    49            Private        218676   Some-college             10   
28029   32   Self-emp-not-inc        189461        HS-grad              9   
3428    46       Self-emp-inc        284799   Some-college             10   
16539   36            Private        226947      Bachelors             13   

            marital-status          occupation    relationship  \
29733   Married-civ-spouse        Adm-clerical         Husband   
235     Married-civ-spouse       Other-service         Husband   
12172        Never-married   Machine-op-inspct   Not-in-family   
5192    Married-civ-spouse     Exec-managerial         Husband   
32511        Never-married        Adm-clerical       Own-child   
...                    ...                 ...             ...   
17534             Divorced    Transport-moving   Not-in-family   
9495    Married-civ-spouse   Machine-op-inspct         Husband   
28029        Never-married       Other-service   Not-in-family   
3428    Married-civ-spouse               Sales         Husband   
16539   Married-civ-spouse      Prof-specialty         Husband   

                      race      sex  capital-gain  capital-loos  \
29733                White     Male             0          1902   
235                  White     Male             0             0   
12172   Asian-Pac-Islander     Male             0             0   
5192                 White     Male         15024             0   
32511                Black   Female             0             0   
...                    ...      ...           ...           ...   
17534                White     Male             0             0   
9495                 White     Male             0             0   
28029                White   Female             0             0   
3428                 White     Male             0             0   
16539                White     Male             0             0   

       hour-per-week  native-country  income  
29733             45   United-States    >50K  
235               40   United-States   <=50K  
12172             40               ?   <=50K  
5192              45   United-States    >50K  
32511             40   United-States   <=50K  
...              ...             ...     ...  
17534             60   United-States   <=50K  
9495              43   United-States   <=50K  
28029             60   United-States   <=50K  
3428              60   United-States   <=50K  
16539             40   United-States    >50K  

[100 rows x 15 columns]

# ----------------------------------------------------------------------

# Amostragem Sistemática

Quando precisamos gerar amostras aleatórias com maior nível de controle é comum o uso de funções auxiliares, para isso irei utilizar a biblioteca <b>numpy</b> do Python.

<hr>

### NUMPY - ARANGE

Lembrando que se houver dúvidas sobre a função <b>arange</b> segue link para a documentação:
<br>
https://numpy.org/doc/stable/reference/generated/numpy.arange.html

<br>

In [8]:
def amostragem_sistematica(dataset, n, r_seed=False):
    # determinando a quantidade de itens para salto entre os indices do dataset
    tam_dataset = len(dataset) // n
    
    # se o valor de r_seed estiver como True a função "seed" manterá os mesmos indices a cada chamada.
    if r_seed:
        random.seed(1)
    
    # definição do inicio e fim do range de valores para a amostra.
    n_rand = random.randint(0, tam_dataset)
    
    indices_amostra_sist = np.arange(n_rand, len(dataset), step=tam_dataset)
    amostra_dataset = dataset.iloc[indices_amostra_sist]
    return amostra_dataset

In [ ]:
amostra_sist = amostragem_sistematica(dataset, 100, True)
amostra_sist

### Algumas dicas
<hr>
1) por quê utilizar 2 barras na divisão?<br>
- a barra de divisão adicional servirá para retornar um valor inteiro (int).
<br><br>
veja o código a seguir, temos 2 exemplos de divisão:<br>
<br>
<b>uma barra:</b><br>
tam_dataset = <b>10 / 3</b> ---> resultado 3.3333...

<b>duas barras:</b><br>
tam_dataset = <b>10 // 3</b> ---> resultado 3<br>

Para este caso de uso será necessário um valor <b>int</b> (tipo inteiro), então usarei 2 barras. No entanto, existem outros métodos que realizam importantes formtações e regras matemáticas, veja na documentação:
https://docs.python.org/3/library/functions.html

<hr>
2) apesar de algunas padronizações na programação, escrever códigos de fácil entendimento nem sempre é uma tarefa fácil mas é uma boa prática declarar variaveis com nomes curtos e sucintos, por exemplo:

- <b>tam_dataset</b>
- <b>indices_amostra_sist</b>

Mesmo encurtando algumas palavras e separando pelo caractere "_" o próprio nome das variáveis já nos indicas quais valores iremos armazenar.
<hr>
<b>3)</b> quando necessário podemos comentar partes do código e informar alguma característica relevante usando outro caractere especial "#".


#### exemplo de como utilizar comentários:
<pre>
#aqui é um comentário de exemplo e servirá para passar alguma orientação ao leitor em "trechos" do código.
</pre>
#### outro exemplo, mais objetivo e informativo:
<pre>
#a seguir irei utilizar duas barras (//) para dividir valores, pois preciso de um resultado do tipo <b>int</b>. 
#este método serve para retornar o valor inteiro em uma operação matemática simples.
#se eu utilizar apenas uma barra (/) poderá retornar um valor decimal.
operacao_div = 10 // 3
</pre>

Neste exemplo de código, o intepretador vai executar apenas a linha com a instrução "<b>operacao_div = 10 // 3</b>", as linhas de código com "#" serão ignoradas.

<br>


# ----------------------------------------------------------------------

# Amostragem Por Grupos


O objetivo desse tipo de amostragem é adicionar caracteristicas para cada elemento, assim é possível obter maior nivel de profundidade na análise.

- Para o exemplo de amostras por agrupamento irei contruir a lógica separadamente para explicar alguns conceitos importantes sobre <b>checagem de valores</b>, após isso irei passar a lógica para uma função.

## tipo de uso para "grupos"

O uso mais comum desse tipo de amostragem (por grupos) é selecionar um grupo randomicamente e explorar os dados desse grupo de forma isolada.

### preparo da amostra
- 1) definir quais os tipos de grupos serão criados (quantidade de grupos);
- 2) selecionar randomicamente um grupo do dataset;

No o código a seguir, criarei uma nova função para gerar os grupos do dataset.

- <b>id_grupo</b>: armazena o do grupo;
- <b>cont_aux</b>: contagem de grupos;
- <b>grupos_dataset</b>: armazena os grupos do dataset;
- <b>quantidade_grupos</b>: quantidade de elementos no dataset dividido por 10.
<br><br>
- utilização do <b>loop for</b> possibilita atribuir valores dinamicamente as variavéis declaradas acima e controlar a quantidade de grupos.<br>
- a função do pandas "iterrows" auxilia o <b>loop for</b> a percorrer todas as linhas do dataset.

In [9]:
id_grupo = 0
cont_aux = 0
grupos_dataset = list()

quantidade_grupos = len(dataset) // 10

for _ in dataset.iterrows():
    grupos_dataset.append(id_grupo)
    cont_aux += 1
    if cont_aux > quantidade_grupos:
        cont_aux = 0
        id_grupo += 1

# Checagem de valores

Para garantir a qualidade da modelagem dos dados é importante realizar a verificação dos valores que serão utilizados junto ao dataset original. O numpy contem algumas funções para checagem rápida de arrays.

# usando numpy para checagem de valores

### unique
- A função <b>unique</b> retorna um array com valores unicos de uma lista.
    - pode ser retornado apenas os valores unicos;
    - tem a opção de retornar um array com quantidade de registros de cada valor unico. 
<pre>
    # apenas valores unicos
    np.unique(grupos_dataset)

    # resumo dos valores unicos
    np.unique(grupos_dataset, return_counts=True)
</pre>

A saída esperada é um array com grupos de 0 a 10.<br>

### shape
- A função <b>shape</b> retorna um array com valores unicos de uma lista.
<pre>
    # checando o shepe da lista de grupos (total de elementos da lista).
    np.shape(grupos_dataset)
    
    # checando o shepe do dataset (total de elementos do dataset).
    dataset.shape
</pre>


A saída esperada é que ambos os valores do "shape" devem serem iguais a 32561.<br>

### Um ponto de atenção

Criei uma lista de grupos aonde não fará diferença se o último grupo terá uma quantidade menor de valores, vale ressaltar que isso não se enquadra em todas as análises. Por isso a importância de reaizar a verificação dos dados complementares.

### checando a quantidade de grupos criados:

In [10]:
# UNIQUE 
print("UNIQUE: ", np.unique(grupos_dataset, return_counts=True))

# SHAPE 
print("\nSHAPE: ", np.shape(grupos_dataset), dataset.shape)

UNIQUE:  (array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]), array([3257, 3257, 3257, 3257, 3257, 3257, 3257, 3257, 3257, 3248],
      dtype=int64))

SHAPE:  (32561,) (32561, 15)


# Anexando os grupos ao dataset

As quantidades de elementos na lista de grupo devem ser iguais a quantidade de elementos do dataset, caso seja diferente será retornar um erro.

Para anexar a nova coluna ao dataset basta "chamar" o dataset e usar a instrução ["nome_da_nova_coluna"] atribuindo o valor da variavel que contem a lista de grupos "grupos_dataset".

In [11]:
dataset["grupo"] = grupos_dataset

### Conferindo os valores do dataset com a nova coluna "grupo"

In [12]:
print(dataset["grupo"].value_counts())

0    3257
1    3257
2    3257
3    3257
4    3257
5    3257
6    3257
7    3257
8    3257
9    3248
Name: grupo, dtype: int64


# Selecionando um grupo randomicamente

In [13]:
# para persistir o mesmo valor use "random.seed(1)"
# este comando gera um valor inteiro de 0 a 9
selecao_grupo_rand = random.randint(0, 9)

# este comando filtra o valor gerado no dataset
df_agrupamento = dataset[dataset["grupo"]==selecao_grupo_rand]
df_agrupamento

age workclass  final-weight      education  education-num  \
9771    24   Private        168997   Some-college             10   
9772    39   Private        168894   Some-college             10   
9773    21   Private        149809     Assoc-acdm             12   
9774    34   Private        344073        HS-grad              9   
9775    22   Private        416165   Some-college             10   
...    ...       ...           ...            ...            ...   
13023   52   Private         42924           12th              8   
13024   43   Private        167599           11th              7   
13025   84         ?        368925        5th-6th              3   
13026   79         ?        100881     Assoc-acdm             12   
13027   35   Private         52738   Some-college             10   

            marital-status          occupation    relationship    race  \
9771    Married-civ-spouse        Craft-repair         Husband   White   
9772         Never-married        Adm-clerical   Not-in-family   White   
9773         Never-married               Sales       Own-child   White   
9774             Separated        Adm-clerical   Not-in-family   White   
9775         Never-married               Sales       Unmarried   White   
...                    ...                 ...             ...     ...   
13023   Married-civ-spouse        Craft-repair         Husband   White   
13024   Married-civ-spouse   Handlers-cleaners            Wife   White   
13025              Widowed                   ?   Not-in-family   White   
13026   Married-civ-spouse                   ?            Wife   White   
13027   Married-civ-spouse   Machine-op-inspct         Husband   White   

           sex  capital-gain  capital-loos  hour-per-week  native-country  \
9771      Male             0             0             40   United-States   
9772    Female             0             0             40   United-States   
9773      Male             0             0             40   United-States   
9774      Male             0             0             40   United-States   
9775    Female             0             0             32   United-States   
...        ...           ...           ...            ...             ...   
13023     Male             0             0             45   United-States   
13024   Female             0             0             25   United-States   
13025     Male             0             0             15   United-States   
13026   Female             0             0              2   United-States   
13027     Male             0             0             46   United-States   

       income  grupo  
9771    <=50K      3  
9772    <=50K      3  
9773    <=50K      3  
9774     >50K      3  
9775    <=50K      3  
...       ...    ...  
13023   <=50K      3  
13024   <=50K      3  
13025   <=50K      3  
13026    >50K      3  
13027   <=50K      3  

[3257 rows x 16 columns]

# Grupo selecionado com sucesso

A variável df_agrupamento agora armazena um grupo gerando randomicamente com o auxilio da biblioteca random.

Lembre-se de utilizar a função random.seed(1) caso precise gerar os mesmos valores repetidamente.


# Convertendo os métodos de criação de grupos para função

Lembrando que uma das boas práticas de codificação é manter sempre seu código organizado, crie instruções dentro de funções para serem legiveis e reutilizaveis.

In [14]:
def amostragem_por_grupos(dataset, n_grupos):
    id_grupo = 0
    cont_aux = 0
    grupos_dataset = list()

    quantidade_grupos = len(dataset) // n_grupos

    for _ in dataset.iterrows():
        grupos_dataset.append(id_grupo)
        cont_aux += 1
        if cont_aux > quantidade_grupos:
            cont_aux = 0
            id_grupo += 1
    
    dataset["grupo"] = grupos_dataset
    selecao_grupo_rand = random.randint(0, n_grupos)
    
    df_agrupamento = dataset[dataset["grupo"]==selecao_grupo_rand]
    return df_agrupamento

In [15]:
df_agrupamento = amostragem_por_grupos(dataset, 22500)
df_agrupamento

age          workclass  final-weight education  education-num  \
2920   63   Self-emp-not-inc         26904   HS-grad              9   
2921   23            Private        238917   7th-8th              4   

           marital-status        occupation     relationship    race    sex  \
2920   Married-civ-spouse   Farming-fishing          Husband   White   Male   
2921        Never-married      Craft-repair   Other-relative   White   Male   

      capital-gain  capital-loos  hour-per-week  native-country  income  grupo  
2920             0             0             98   United-States   <=50K   1460  
2921             0             0             36   United-States   <=50K   1460

# Escalabilidade do código

Perceba que ao criar a função "<b>amostragem_por_grupos</b>" é possível com apenas uma linha de código gerar diferentes quantidade de grupos.

No código fora da função eu havia definido um valor fixo "10", agora posso reutilizar a função "amostragem_por_grupos" e passar os argumentos "dataset" e os valores que achar necessário em "n_grupos" assim gerando diferentes amostras.

São pequenos ajustes que farão bastante diferença na legibilidade e performance da sua aplicação.

# Amostra Estratificada

Este tipo de amostra é bastante utilizada quando precisamos analisar um percentual de uma população.

- exemplo: em uma população há 100 pessoas, precisamos analisar apenas os dados das mulheres (de forma simplificada temos que existam apenas duas opções: mulheres e homens).
- supondo que existam 60 mulheres e 40 homens na amostra.

Percebam que a quantidade de mulheres é maior do que a quantidade de homens, para amostras estratificadas devemos selecionar proporcionalmente a quantidade de elementos de forma que sejam equivalentes para ambos os conjuntos de dados, ou seja, é necessário obter uma amostra em percentual.

### Queremos analisar apenas 10% das mulheres e 10% dos homens.

#### mulheres
- 60 * 10 / 100 = 6,0

#### homens
- 40 * 10 / 100 = 4,0

Assim temos que deverá ser analisado <b>6 mulheres</b> e <b>4 homens</b>.

#### Em caso de valores quebrados, exemplo: 6,4.
Podemos aplicar a regra de arredondamento para menor ou maior ficando 6, caso seja >= 6,5 poderiamos arredondar para 7.

### Importante saber

Esse método serve para selecionar um subconjunto de dados, a análise deverá ser complementada por uma tecnica auxiliar como por exemplo, seleção randomica simples.


# vamos para implementação da amostragem estratificada

### instalação da biblioteca sklearn

#### instalação no gerenciador de pacotes pip
<pre>
pip install -U scikit-learn
</pre>

#### instalação no gerenciador de pacotes conda
<pre>
conda create -n sklearn-env -c conda-forge scikit-learn
conda activate sklearn-env
</pre>


### para mais plataformas, acesse a documentação de instalação:
https://scikit-learn.org/stable/install.html

# Importação da função StratifiedShuffleSplit



In [16]:
from sklearn.model_selection import StratifiedShuffleSplit

# Renda - Amostra Estratificada

### Vamos analisar a renda da população

Aplicando o método <b>value_counts</b> na coluna <b>income</b> chegaremos a um resumo da renda em nosso dataset.

Resumo da renda:
<pre>
dataset["income"].value_counts()

# saída:
# <=50K    24720
# >50K      7841
</pre>

- total de pessoas que ganham <= 50 mil:..24720 --> equivale a 75,91%;
- total de pessoas que ganham >  50 mil:......7841 -->  equivale a 24,09%;

### objetivo: gerar uma amostra da população para calcular a representatividade de renda

Veja que a quantidade de pessoas com renda <= 50k é desproporcional, sabendo disso fica evidente que se utilizar a amostra randomica pode haver a influência maior de seleção para elementos que possuam renda <=50k e comprometer a análise.

Vamos aplicar a amostra estratificada para seguir com a análise.

In [27]:
dataset["income"].value_counts()

 <=50K    24720
 >50K      7841
Name: income, dtype: int64

In [32]:
# percentuais de renda
7841 / len(dataset), 24720 / len(dataset)

(0.2408095574460244, 0.7591904425539756)

# amostra de treino e teste

Preparando amostras de forma inteligente...

Atualmente na área de dados é comum o uso de algoritimos de estatística e machine learning, já existem divesas bibliotecas desenvolvidas em Python com o objetivo de auxiliar na exploração de dados.

Um método bastante comum é a separação de amostras para treinar a máquina, após parametrizado é aplicado testes para validar os resultados.

### exemplos para dividir as amostras
- amostra para treino: 90%;
- amostra para teste: 10%.

A variavel <b>percentual_base</b> indicará o tamanho da amostra, no caso "0.1" representa 10% da base, será utilizado a função do sklearn <b>StratifiedShuffleSplit</b> para divisão da amostra treino e teste.

- percentual_base: tamanho da amostra;
- divisao: variavel que armazenará a função StratifiedShuffleSplit;
- df_x: deve representar 90% da base (amostra para treino);
- <b>df_y</b>: deve representar 10% da base <b>(amostra para teste)</b>;


In [46]:
percentual_base = 0.1
divisao = StratifiedShuffleSplit(test_size=percentual_base)

for x, y in divisao.split(dataset, dataset["income"]):
    df_x = dataset.iloc[x]
    df_y = dataset.iloc[y]

df_x.shape, df_y.shape

((29304, 16), (3257, 16))

Perceba a facilidade, com um loop for simples é possível definir os valores de teste e treino.

Mas podemos melhorar o código e para isso irei criar uma função chamada <b>amostra_estratificada</b> e deixar o código reutilizavél.

In [63]:
def amostra_estratificada(base, test_size, coluna):
    percentual_base = test_size / len(dataset)
    divisao = StratifiedShuffleSplit(test_size=test_size)

    for x, y in divisao.split(dataset, dataset[coluna]):
        df_x = dataset.iloc[x]
        df_y = dataset.iloc[y]
    return df_x, df_y

# Uso da função de amostra estratificada

Pequenas mudanças foram realizada para deixar o código limpo e reutilizavel.

Basta informar os dados para função:
- base: nosso dataset;
- test_size: quantidade de registros para a amostra;
- coluna: nome da coluna que será a referência para função "split".

In [64]:
amostra_estratif = amostra_estratificada(dataset, 100, "income")

In [65]:
amostra_treino = amostra_estratif[0]
amostra_teste = amostra_estratif[1]

amostra_treino.shape, amostra_teste.shape

((32461, 16), (100, 16))

In [75]:
amostra_teste

age          workclass  final-weight     education  education-num  \
6954    47          Local-gov        121124       Masters             14   
7989    47          Local-gov        174533     Assoc-voc             11   
22385   47            Private        246739     Bachelors             13   
14193   17            Private        166290           9th              5   
17717   30                  ?         33811     Bachelors             13   
...    ...                ...           ...           ...            ...   
29480   24                  ?        138504       HS-grad              9   
26724   59   Self-emp-not-inc        148626          10th              6   
23955   31            Private        144064       HS-grad              9   
2672    38            Private        284166   Prof-school             15   
21809   41   Self-emp-not-inc         26598       HS-grad              9   

            marital-status        occupation    relationship    race      sex  \
6954    Married-civ-spouse    Prof-specialty         Husband   White     Male   
7989    Married-civ-spouse   Exec-managerial         Husband   White     Male   
22385   Married-civ-spouse   Exec-managerial         Husband   White     Male   
14193        Never-married     Other-service       Own-child   White   Female   
17717   Married-civ-spouse                 ?            Wife   Other   Female   
...                    ...               ...             ...     ...      ...   
29480            Separated                 ?       Unmarried   Black   Female   
26724   Married-civ-spouse   Farming-fishing         Husband   White     Male   
23955   Married-civ-spouse      Craft-repair         Husband   White     Male   
2672         Never-married    Prof-specialty   Not-in-family   White     Male   
21809             Divorced      Craft-repair   Not-in-family   White     Male   

       capital-gain  capital-loos  hour-per-week  native-country  income  \
6954              0             0             35   United-States    >50K   
7989              0             0             40   United-States    >50K   
22385         99999             0             55   United-States    >50K   
14193             0             0             20   United-States   <=50K   
17717             0             0             40          Taiwan    >50K   
...             ...           ...            ...             ...     ...   
29480             0             0             37   United-States   <=50K   
26724             0             0             40   United-States   <=50K   
23955             0             0             48   United-States    >50K   
2672              0             0             50   United-States   <=50K   
21809             0             0             40   United-States   <=50K   

       grupo  
6954    3477  
7989    3994  
22385  11192  
14193   7096  
17717   8858  
...      ...  
29480  14740  
26724  13362  
23955  11977  
2672    1336  
21809  10904  

[100 rows x 16 columns]

# Validando valores da amostra estratificada

Instruimos a função a gerar 100 registros de amostra estratificada, agora precisamos validar os valores gerados.

Já sabemos como resumir os valores de um dataset, basta informar a coluna e usar a função <b>value_counts</b>.

<b>relembrando que...</b><br>
a base de treino serve apenas para usarmos em treinos de algoritimos.

Para aplicação dos métodos <b>considerem sempre a base de teste</b>.


In [72]:
amostra_teste["income"].value_counts()

 <=50K    76
 >50K     24
Name: income, dtype: int64

Como podemos observar, os resultados são equivalentes ao percentual geral do dataset.

De 100 registros, temos que:
- elementos com renda <=50K:....76 registros --> equivale a 76%;
- elementos com renda >50K:......24 registros --> equivale a 24%;

Utilizando o sklearn foi possível separar uma amostra de 100 registros para análise de subgrupos de renda de forma equivalente, além disso a função <b>amostra_estratificada</b> proporciona reutilização do código de forma simplificada, podemos escolher quantidades diferentes e até mesmo outras colunas para gerar mais amostras estratificadas.